WooCommerce/Classes/Authentication/Epilogue/SwitchStoreUseCase.swift

```diff
@@ -0,0 +1,89 @@
+import UIKit
+import Yosemite
+
+
+/// Simplifies and decouples the store picker from the caller
+///
+final class SwitchStoreUseCase {
+
+    private let stores: StoresManager
+
+    init(stores: StoresManager) {
+        self.stores = stores
+    }
+
+    /// A static method which allows easily to switch store. The boolean argument in `onCompletion` indicates that the site was changed.
+    ///
+    func switchStore(with storeID: Int64, onCompletion: @escaping (Bool) -> Void) {
+        guard storeID != stores.sessionManager.defaultStoreID else {
+            onCompletion(false)
+            return
+        }
+
+        logOutOfCurrentStore { [weak self] in
```

Comment:
I'm sorry I didn't notice this before. But using `weak self` here prematurely deallocates `SwitchStoreUseCase`. The result is that when receiving a push notification from a non-active site, no site-switching happens and the order is loaded while on the wrong site. 

### Steps to reproduce

1. Following testing steps in PR description.
2. Submit an order for the old site. 
3. Tap on the order push notification.
4. Observe that the `Notice` being is shown mentions _the old site_ and not the current one. 
5. Observe the order is loaded but the app is currently on the wrong site. 

### Why

This happens because the code for using the `SwitchStoreUseCase` does not **retain** it: 

https://github.com/woocommerce/woocommerce-ios/blob/82ba43d31c9e7997fc205890788e29bfebc295ff/WooCommerce/Classes/ViewRelated/MainTabBarController.swift#L291

So the callback in `logOutOfCurrentStore` is called but with a non-existent `self`. Which means `finalizeStoreSelection` is not called.

It works in the store picker because the `UseCase` is retained there:

https://github.com/woocommerce/woocommerce-ios/blob/82ba43d31c9e7997fc205890788e29bfebc295ff/WooCommerce/Classes/Authentication/Epilogue/StorePickerCoordinator.swift#L21

### The fix

I tested it and this can be fixed by keeping `self` as `strong` in this line. 

```suggestion
        logOutOfCurrentStore { 
```

I believe this fine. Just like how we use callbacks in Networking or other APIs, it should be the responsibility of the **call site** to check for `weak self`. Something like:

```swift
            SwitchStoreUseCase(stores: ServiceLocator.stores).switchStore(with: Int64(siteID)) { [weak self] siteChanged in
                guard let self = self else {
                    return
                }

                presentNotificationDetails(for: note)

                if siteChanged {
                    SwitchStoreNoticePresenter.presentStoreSwitchedNotice(stores: ServiceLocator.stores, configuration: .switchingStores)
                }
            }
```

Note that this code isn't going to work for `MainTabBarController` because that's in a `static func`. 🙃

Reply:
Really great discovery @shiki 👏 This should fix it, as suggest by you: 2c63a2627d300cb7febc2b48b5baaa0fef42df86 
I also added a comment, just to remember why we shouldn't use `[weak self]` here.